src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java

```diff
@@ -81,4 +81,113 @@ public SearchResultPageDTO<LearningGoal> getAllOnPageWithSize(final PageableSear
+            }
+
+            public boolean isBeingVisited() {
+                return beingVisited;
+            }
+
+            public void setBeingVisited(boolean beingVisited) {
+                this.beingVisited = beingVisited;
+            }
+
+            public boolean isVisited() {
+                return visited;
+            }
+
+            public void setVisited(boolean visited) {
+                this.visited = visited;
+            }
+        }
+
+        // Inner class Graph is only used in this method for cycle detection
+        class Graph {
+
+            private final List<Vertex> vertices;
+
+            public Graph() {
+                this.vertices = new ArrayList<>();
+            }
+
+            public void addVertex(Vertex vertex) {
+                this.vertices.add(vertex);
+            }
+
+            public void addEdge(Vertex from, Vertex to) {
+                from.addNeighbor(to);
+            }
+
+            // Checks all vertices of the graph if they are part of a cycle.
+            // This is necessary because otherwise we would not traverse the complete graph if it is not connected
+            public boolean hasCycle() {
+                for (Vertex vertex : vertices) {
+                    if (!vertex.isVisited() && vertexIsPartOfCycle(vertex)) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            public boolean vertexIsPartOfCycle(Vertex sourceVertex) {
+                sourceVertex.setBeingVisited(true);
+                for (Vertex neighbor : sourceVertex.getAdjacencyList()) {
+                    if (neighbor.isBeingVisited()) {
+                        // backward edge exists
+                        return true;
+                    }
+                    else if (!neighbor.isVisited() && vertexIsPartOfCycle(neighbor)) {
+                        return true;
+                    }
+                }
+                sourceVertex.setBeingVisited(false);
+                sourceVertex.setVisited(true);
+                return false;
+            }
+        }
+
+        var graph = new Graph();
+        for (LearningGoal learningGoal : learningGoals) {
+            graph.addVertex(new Vertex(learningGoal.getTitle()));
+        }
+        for (LearningGoalRelation relation : relations) {
+            var headVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getHeadLearningGoal().getTitle())).findFirst().orElseThrow();
+            var tailVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getTailLearningGoal().getTitle())).findFirst().orElseThrow();
+            // Only EXTENDS and ASSUMES are included in the generated graph as other relations are no problem if they are circular
+            switch (relation.getType()) {
+                case EXTENDS, ASSUMES -> graph.addEdge(tailVertex, headVertex);
+            }
```

Comment:
I think you also have to consider MATCHES here. Otherwise I could have A matching C and B assuming A and C assuming C, so A -> B -> C=A. Or A matching B and B matching C and C assuming A, so A=B=C and A -> C

You could combine the matching vertices after adding the edges. If one of the combined vertices includes the other in its neighbors, you know the graph has a cycle. Otherwise you merge the two vertices and also merge their neighbor lists. After that you can continue with the normal cycle check after combining all matches.
You could think about switching the adjacency list to a set since it handles potential duplicates from merging automatically and it is faster for checking if a certain vertex is a neighbor.

Reply:
You are right. I also have to consider MATCHES.
Thank you for also providing a good solution!
I added this in https://github.com/ls1intum/Artemis/pull/6400/commits/269861f945982d56ba3d6a93106ecff44cd56e96 and also changed adjacency lists to sets.